lib/ns.c

```diff
@@ -368,54 +468,191 @@ DEFUN (no_ns_netns,
 			continue;
-
 		vty_out(vty, "logical-router %u netns %s\n", ns->ns_id,
 			ns->name);
 		write = 1;
 	}
-
 	return write;
 }
 
-/* Initialize NS module. */
+DEFUN_NOSH (ns_netns,
+	    ns_netns_cmd,
+	    "netns NAME",
+	    "Attach VRF to a Namespace\n"
+	    "The file name in " NS_RUN_DIR ", or a full pathname\n")
+{
+	int idx_name = 1;
+	char *pathname = ns_netns_pathname(vty, argv[idx_name]->arg);
+
+	VTY_DECLVAR_CONTEXT(vrf, vrf);
+
+	if (!pathname)
+		return CMD_WARNING_CONFIG_FAILED;
+	return ns_handler_create(vty, vrf, pathname, NS_UNKNOWN);
+}
+
+DEFUN (no_ns_netns,
+	no_ns_netns_cmd,
+	"no netns [NAME]",
+	NO_STR
+	"Detach VRF from a Namespace\n"
+	"The file name in " NS_RUN_DIR ", or a full pathname\n")
+{
+	struct ns *ns = NULL;
+
+	VTY_DECLVAR_CONTEXT(vrf, vrf);
+
+	if (!vrf_is_backend_netns()) {
+		vty_out(vty, "VRF backend is not Netns. Aborting\n");
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+	if (!vrf->ns_ctxt) {
+		vty_out(vty, "VRF %s(%u) is not configured with NetNS\n",
+			vrf->name, vrf->vrf_id);
+		return CMD_WARNING_CONFIG_FAILED;
+	}
+
+	ns = (struct ns *)vrf->ns_ctxt;
+
+	ns->vrf_ctxt = NULL;
+	vrf_disable(vrf);
+	vrf_update_vrf_id((vrf_id_t) VRF_UNKNOWN, vrf);
+	RB_REMOVE(ns_head, &ns_tree, ns);
+	ns->ns_id = NS_UNKNOWN;
+	RB_INSERT(ns_head, &ns_tree, ns);
+	ns_delete(ns);
+	vrf->ns_ctxt = NULL;
+	return CMD_SUCCESS;
+}
+
 void ns_init(void)
 {
-	struct ns *default_ns;
+#ifdef HAVE_NETNS
+	if (have_netns_enabled < 0) {
+		ns_default_ns_fd = open(NS_DEFAULT_NAME, O_RDONLY);
+		return;
+	}
+#endif /* HAVE_NETNS */
+	ns_default_ns_fd = -1;
+	default_ns = NULL;
+}
 
-	/* The default NS always exists. */
-	default_ns = ns_get(NS_DEFAULT);
+/* Initialize NS module. */
+void ns_init_zebra(ns_id_t default_ns_id)
+{
+	int fd;
+
+	ns_init();
+	default_ns = ns_get(default_ns_id);
 	if (!default_ns) {
 		zlog_err("ns_init: failed to create the default NS!");
 		exit(1);
 	}
-
+	if (have_netns()) {
+		fd = open(NS_DEFAULT_NAME, O_RDONLY);
+		default_ns->fd = fd;
+	}
+	ns_current_ns_fd = -1;
 	/* Set the default NS name. */
 	default_ns->name = XSTRDUP(MTYPE_NS_NAME, NS_DEFAULT_NAME);
+	zlog_info("ns_init: default NSID is %u", default_ns->ns_id);
 
 	/* Enable the default NS. */
 	if (!ns_enable(default_ns)) {
 		zlog_err("ns_init: failed to enable the default NS!");
 		exit(1);
 	}
 
-	if (have_netns()) {
+	if (have_netns() && !vrf_is_backend_netns()) {
 		/* Install NS commands. */
-		install_node(&ns_node, ns_config_write);
-		install_element(CONFIG_NODE, &ns_netns_cmd);
-		install_element(CONFIG_NODE, &no_ns_netns_cmd);
+		install_node(&logicalrouter_node, ns_logicalrouter_config_write);
+		install_element(CONFIG_NODE, &ns_logicalrouter_cmd);
+		install_element(CONFIG_NODE, &no_ns_logicalrouter_cmd);
+	}
+}
+
+void ns_cmd_init()
+{
+	if (have_netns() && vrf_is_backend_netns()) {
+		/* Install NS commands. */
+		install_element(VRF_NODE, &ns_netns_cmd);
+		install_element(VRF_NODE, &no_ns_netns_cmd);
```

Comment:
I agree on the first remark to move those commands under vrf.
I diagree on the second comment ( only related to logical router).

Reply:
Ok, but why do you disagree?

Just one silly example to illustrate my point: configure multiple VRFs using the netns backend and then type the "show memory" command. You'll see that zebra allocated memory for several logical routers, which of course is not right. I think we need to implement the right abstractions instead of reusing data structure for multiple purposes. Let's discuss this tomorrow in more detail.

Reply:
I mean, the way you understand ns.c differs from my perspective.
ns.c is not dedicated to logical-router.
maybe you would like to have a split of this file in two: one for logical router, and a kind of ns toolbox backend.
Indeed, as the logical router feature does not work yet, I don't see why I would spend some time to prepare that work.